rust-src/concordium_base/src/web3id/v1.rs

```diff
@@ -0,0 +1,2092 @@
+        map.serialize_entry("context", &self.challenge)?;
+        map.serialize_entry("subjectClaims", &self.subject_claims)?;
+        map.end()
+    }
+}
+
+impl<'de, C: Curve, AttributeType: Attribute<C::Scalar> + DeserializeOwned> serde::Deserialize<'de>
+    for RequestV1<C, AttributeType>
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let mut value = serde_json::Value::deserialize(deserializer)?;
+
+        let result = (|| -> anyhow::Result<Self> {
+            let ty: String = take_field_de(&mut value, "type")?;
+            ensure!(
+                ty == CONCORDIUM_REQUEST_TYPE,
+                "expected type {}",
+                CONCORDIUM_REQUEST_TYPE
+            );
+
+            let challenge = take_field_de(&mut value, "context")?;
+            let credential_statements = take_field_de(&mut value, "subjectClaims")?;
+
+            Ok(Self {
+                challenge,
+                subject_claims: credential_statements,
+            })
+        })();
+
+        result.map_err(|err| D::Error::custom(format!("{:#}", err)))
+    }
+}
+
+/// Private inputs for an account credential proof.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct AccountCredentialProofPrivateInputs<'a, C: Curve, AttributeType> {
+    /// Issuer of the identity credentials used to deploy the account credentials
+    pub issuer: IpIdentity,
+    /// The attribute values that are committed to in the account credentials
+    pub attribute_values: &'a BTreeMap<AttributeTag, AttributeType>,
+    /// The randomness of the attribute commitments in the account credentials
+    pub attribute_randomness: &'a BTreeMap<AttributeTag, pedersen_commitment::Randomness<C>>,
+}
+
+/// Private inputs for an identity credential proof.
+pub struct IdentityCredentialProofPrivateInputs<
+    'a,
+    P: Pairing,
+    C: Curve<Scalar = P::ScalarField>,
+    AttributeType,
+> {
+    /// Identity provider data
+    pub ip_context: IpContextOnly<'a, P, C>,
+    /// Identity object. Together with `id_object_use_data`, it constitutes the identity credentials
+    pub id_object: &'a dyn HasIdentityObjectFields<P, C, AttributeType>,
+    /// Identity credentials
+    pub id_object_use_data: &'a IdObjectUseData<P, C>,
+}
+
+/// The additional private inputs (mostly secrets), needed to prove the claims
+/// in a [request](RequestV1).
+pub enum CredentialProofPrivateInputs<
+    'a,
+    P: Pairing,
+    C: Curve<Scalar = P::ScalarField>,
+    AttributeType,
+> {
+    /// Inputs are for an account credential derived from an identity issued by an
+    /// identity provider.
+    Account(AccountCredentialProofPrivateInputs<'a, C, AttributeType>),
+    /// Inputs are for an identity credential issued by an identity provider.
+    Identity(IdentityCredentialProofPrivateInputs<'a, P, C, AttributeType>),
+}
+
+/// An owned version of [`CredentialVerificationMaterial`] that can be deserialized.
```

Comment:
Shouldn't for the following types also the `serde::Serialize` implemented to be complet?